### ISSUE TITLE
Upper-bounded generic type relaxes bigint restriction

### ISSUE THREAD

**Issue by theonlypwner** [microsoft/TypeScript#49907]
# Bug Report

### 🔎 Search Terms

bug extends number bigint

### 🕗 Version & Regression Information

- This was reported in #44578 and occurs in playground on v4.74 and v4.8.0-beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAdgrgWwEYQE4HlUCECWBzHGYKAXlkRVSgB8ol9DgBuAWACh2AzOGAY2BwB7GFDwRgAaQggAFAGtpALnLI0mXASIBKKAG92UQ1AD0xqAAFgAZwC0EAB6R+d1KkFVAKOQGjAagUhWNiMTM0tbBydgFzdPb0MABigfKH9AgF92Lh5+IRExSWkAJgAeAGUoB2AIGAATKxVKdQYiAD55JShSnX0go1MLaztHCGc0GKhAGXIoKwALQTgAGxqK13cAGjo4YmAZ6F5BBDAcBbQoGsEIephBYDik1Lv+sKHI6PcoL16EpJTpdKA)

[Playground link with original code from #44578](https://www.typescriptlang.org/play?ssl=11&ssc=2&pln=1&pc=1#code/C4TwDgpgBAdgrgWwEYQE4HlUCECWBzHGYKAXlkRVSgB8ol9DgBuAWACh2AzOGAY2BwB7GFDwRgAaQggAFAGtpALnLI0mXASIBKKAG92UQ1ADUCkEyMB6S1EAo5FADOAC0FwANgBMoaVINQAaKABDGC9gJ2heQQQwHDc0KA9BCAcDIwAGEygzCyhrO0cXdy8fP0CQsIioKJi4hKSU9gBfdi4efiERMUlpACYAHgBlbwAPYAhQhxVKdQYiAD55JShBnX02IxMcw3zAGXJC109vVF8qQKQ4YnDI6Nj4qgapmEFgNMNM42zpXLybe2dDiUTmVgqEoNdqrc6g9kqk2E0gA)

### 💻 Code

```ts
type numberOrBigint = number | bigint;

function getKey(key: numberOrBigint) {
    // @ts-expect-error ✔
    +key;
    // @ts-expect-error ✔
    0 + key;
}

function getKey2<S extends numberOrBigint>(key: S) {
    // @ts-expect-error ❌ should error, but the compiler does not
    +key;
    // @ts-expect-error ✔
    0 + key;
}
```

### 🙁 Actual behavior

`+key` is not an error in `getKey2`.

### 🙂 Expected behavior

It should be an error to be consistent with `getKey`, as `S` could be `numberOrBigint` instead of a more specific type.
